src/modes/ModeEstimation.jl

```diff
@@ -0,0 +1,284 @@
+)
+    r = reshape(vi[vec(vns)], size(vns))
+    return r
+end
+
+"""
+    OptimLogDensity{M<:Model,C<:Context,V<:VarInfo}
+
+A struct that stores the log density function of a `DynamicPPL` model.
+"""
+struct OptimLogDensity{M<:Model,C<:AbstractContext,V<:VarInfo}
+    "A `DynamicPPL.Model` constructed either with the `@model` macro or manually."
+    model::M
+    "A `DynamicPPL.AbstractContext` used to evaluate the model. `LikelihoodContext` or `DefaultContext` are typical for MAP/MLE."
+    context::C
+    "A `DynamicPPL.VarInfo` struct that will be used to update model parameters."
+    vi::V
+end
+
+"""
+    OptimLogDensity(model::Model, context::AbstractContext)
+
+Create a callable `OptimLogDensity` struct that evaluates a model using the given `context`.
+"""
+function OptimLogDensity(model::Model, context::AbstractContext)
+    init = VarInfo(model)
+    DynamicPPL.link!(init, DynamicPPL.SampleFromPrior())
+    return OptimLogDensity(model, context, init)
+end
+
+"""
+    (f::OptimLogDensity)(z)
+
+Evaluate the log joint (with `DefaultContext`) or log likelihood (with `LikelihoodContext`)
+at the array `z`.
+"""
+function (f::OptimLogDensity)(z)
+    spl = DynamicPPL.SampleFromPrior()
+
+    varinfo = DynamicPPL.VarInfo(f.vi, spl, z)
+    f.model(varinfo, spl, f.context)
+    return -DynamicPPL.getlogp(varinfo)
+end
+
+"""
+    ModeResult{
+        V<:NamedArrays.NamedArray, 
+        M<:NamedArrays.NamedArray, 
+        O<:Optim.MultivariateOptimizationResults, 
+        S<:NamedArrays.NamedArray
+    }
+
+A wrapper struct to store various results from a MAP or MLE estimation.
+"""
+struct ModeResult{
+    V<:NamedArrays.NamedArray, 
+    O<:Optim.MultivariateOptimizationResults,
+    M<:OptimLogDensity
+} <: StatsBase.StatisticalModel
+    "A vector with the resulting point estimates."
+    values :: V
+    "The stored Optim.jl results."
+    optim_result :: O
+    "The final log likelihood or log joint, depending on whether `MAP` or `MLE` was run."
+    lp :: Float64
+    "The evaluation function used to calculate the output."
+    f :: M
+end
+#############################
+# Various StatsBase methods #
+#############################
+
+
+
+function Base.show(io::IO, ::MIME"text/plain", m::ModeResult)
+    print(io, "ModeResult with minimized lp of ")
+    Printf.@printf(io, "%.2f", m.lp)
+    println(io)
+    show(io, m.values)
+end
+
+function Base.show(io::IO, m::ModeResult)
+    show(io, m.values.array)
+end
+
+function StatsBase.coeftable(m::ModeResult)
+    # Get columns for coeftable.
+    terms = StatsBase.coefnames(m)
+    estimates = m.values.array[:,1]
+    stderrors = StatsBase.stderror(m)
+    tstats = estimates ./ stderrors
+
+    StatsBase.CoefTable([estimates, stderrors, tstats], ["estimate", "stderror", "tstat"], terms)
+end
+
+function StatsBase.informationmatrix(m::ModeResult; hessian_function=ForwardDiff.hessian, kwargs...)
+    # Calculate Hessian and information matrix.
+    varnames = StatsBase.coefnames(m)
```

Comment:
Should probably check the vi is invlinked here. If not, then run `invlink!`.

Reply:
What's the quick, idiomatic way to check whether the whole VI is linked?

Reply:
islinked(vi, SampleFromPrior())` should be good enough. It will only check the first varname of each symbol though.

Reply:
Okay, I made a quick work around to get this to work with `SampleFromPrior`, since `islinked` only works with `Sampler`. I opened a PR to fix this here: https://github.com/TuringLang/DynamicPPL.jl/pull/124

For right now, I think it's fine to use my quick workaround and replace it later if/when the DPPL PR goes through.